internal/login/login_server_routes.go

```diff
@@ -0,0 +1,178 @@
+package login
+
+import (
+	"fmt"
+	"log/slog"
+	"net/http"
+
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/models"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/sessions"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/utils"
+	"github.com/labstack/echo/v4"
+)
+
+// GetLogin is a handler for the initiation of a authorization code flow login for Renku
+func (l *LoginServer) GetLogin(c echo.Context, params GetLoginParams) error {
+	session, err := l.sessions.Create(c)
+	if err != nil {
+		return err
+	}
+	// Check redirect parameters
+	var appRedirectURL string
+	if params.RedirectUrl != nil && *params.RedirectUrl != "" {
+		appRedirectURL = *params.RedirectUrl
+	} else {
+		appRedirectURL = l.config.RenkuBaseURL.String()
+	}
+	session.LoginRedirectURL = appRedirectURL
+	// Check provider IDs requested for login
+	var loginSequence models.SerializableStringSlice
+	if params.ProviderId != nil && len(*params.ProviderId) > 0 {
+		loginSequence = *params.ProviderId
+	} else {
+		loginSequence = defaultLoginSequence[:]
+	}
+	session.LoginSequence = loginSequence
+	return l.nextAuthStep(c, session)
+}
+
+func (l *LoginServer) GetCallback(c echo.Context, params GetCallbackParams) error {
+	// Load both the regular and the cli session (if present), see which fits
```

Comment:
I think this comment does not apply anymore.

Reply:
Done in b2eed526e1bdc4968bcf6a93eca3d08f1c32d8fb.